phoenix-core-server/src/main/java/org/apache/phoenix/hbase/index/IndexRegionObserver.java

```diff
@@ -1588,10 +1680,36 @@ private List<Mutation> generateOnDupMutations(BatchMutateContext context, Put at
               delete.add(cell);
           }
       }
+
+      if (put != null) {
+          // if put is empty and delete is null, remove the empty put from mutations,
+          // otherwise we should also add empty cell timestamp for update
+          if (put.isEmpty() && delete == null) {
+              mutations.remove(put);
+          } else {
+              addEmptyKVCells(put, delete, tuple);
+          }
+      }
+
       return mutations;
   }
 
-
+    private void addEmptyKVCells(Put put, Delete delete, MultiKeyValueTuple tuple) throws IOException {
+        Set<byte[]> familySet = new HashSet<>();
+        if (!put.isEmpty()) {
+            familySet.addAll(put.getFamilyCellMap().keySet());
+        }
+        if (delete != null) {
+            familySet.addAll(delete.getFamilyCellMap().keySet());
+        }
+        for (byte[] familyBytes : familySet) {
+            Cell emptyKVCell = tuple.getValue(familyBytes,
+                    QueryConstants.ENCODED_EMPTY_COLUMN_BYTES);
```

Comment:
We can't always assume the empty column will be encoded.

Reply:
+1, we need to use `EncodedColumnsUtil#usesEncodedColumnNames(PTable)` with the PTable object to determine whether empty column need to be encoded, else use `EMPTY_COLUMN_BYTES`.

Reply:
Since we don't seem to have PTable object here (unless used by expression evaluation), probably it might be good to send boolean flag (isTableQualifiersEncoded) from client side as Mutation attribute. WDYT @tkhurana?

Reply:
We already serializing the PTable object and re-creating on the server so we can get the encoding information https://github.com/apache/phoenix/blob/master/phoenix-core-server/src/main/java/org/apache/phoenix/hbase/index/IndexRegionObserver.java#L1482

Reply:
I still don't know how to pair the put cells with the PTable object